editquality/utilities/extract_balanced_sample.py

```diff
@@ -0,0 +1,174 @@
+    --revert-window=<secs>      The maximum amount of time to wait for a
+                                revision to be reverted [default: 172800]
+    --rev-reverteds=<path>      The location to write output to.
+                                [default: <stdout>]
+    --trusted-groups=<groups>   User groups that should be considered trusted.
+    --trusted-edits=<num>       Minimum number of edits to be considered
+                                trusted.
+    --verbose                   Prints dots and stuff to stderr
+"""
+import sys
+import docopt
+import mysqltsv
+
+import mwapi
+import mwreverts
+import mwxml
+from mwtypes import Timestamp
+
+
+class Revision(object):
+    """Placeholder"""
+    def __init__(self, id):
+        self.id = id
+        self.status = False
+        self.reason = []
+
+
+class User(object):
+    """Placeholder"""
+    def __init__(self, id, editcount=0, groups=[]):
+        self.id = id
+        self.editcount = editcount
+        self.usergroups = groups
+
+
+def load_user_info(user_name, session):
+    res = session.get(action='query', list='users', ususers=user_name,
+                       usprop='groups|editcount')
+    user_data = res["query"]["users"][0]
+    if 'userid' not in user_data:
+        return User(user_name, 0, [])
+    user = User(user_data["userid"], user_data['editcount'],
+                user_data["groups"])
+    return user
+
+
+def page_data(dump, path):
+    for page in dump:
+        yield page
+
+
+def main(argv=None):
+    args = docopt.docopt(__doc__, argv=argv)
+
+    if args['--rev-reverteds'] == "<stdout>":
+        rev_reverteds = mysqltsv.Writer(sys.stdout)
+    else:
+        rev_reverteds = mysqltsv.Writer(open(args['--rev-reverteds'], "w"))
+
+    revert_radius = int(args['--revert-radius'])
+    revert_window = int(args['--revert-window'])
+
+    if args['--host']:
+        session = mwapi.Session(
+            args['--host'], user_agent="ORES revert labeling utility")
+    else:
+        session = None
+    dumps = args['<dump-file>']
+
+    verbose = args['--verbose']
+    start = args['--start']
+    if start:
+        start = Timestamp(start)
+    end = args['--end']
+    if end:
+        end = Timestamp(end)
+    reverted_only = args['--reverted-only']
+    trusted_groups = args['--trusted-groups']
+    if trusted_groups:
+        trusted_groups = trusted_groups.split(',')
+    min_editcount = args['--trusted-edits']
+    if min_editcount:
+        min_editcount = int(min_editcount)
+
+    run(dumps, session, start, end, revert_radius, revert_window,
+        reverted_only, trusted_groups, min_editcount, rev_reverteds,
+        verbose=verbose)
+
+
+def run(dumps, session, start, end, revert_radius, revert_window,
+        reverted_only, trusted_groups, min_editcount, rev_reverteds,
+        verbose=False):
+
+    user_data = {}
+    for page in mwxml.map(page_data, dumps):
+        detector = mwreverts.Detector()
```

Comment:
Did you test this?  It looks like it won't work.  You need to iterate through the page in the dump_processor function.